### DISCUSSION TITLE
Mac mini M1

### DISCUSSION THREAD

**Discussion by duckdany** [AlexGustafsson/homebridge-wol#178] [Questions]
I have changed from iMac 27" Late 2013 to Mac mini M1 (2020).
From Homebridge terminal shutdown command works fine but on my iPhone not working.
And the ping is not working anymore. I got this error:

[18/10/2021, 23:47:59] [Mac mini] Attempting to poll state of "Mac mini" (192.168.28.12) using "ssh"
[18/10/2021, 23:48:00] [Mac mini] Result of executing ping command for "Mac mini" (192.168.28.12): offline
[18/10/2021, 23:48:00] [Mac mini] Error: Command failed: ssh dani@Daniel-Mac-minije.local 'if [[ $(pmset -g powerstate IODisplayWrangler | tail -1 | cut -c29) -lt 4 ]]; then; exit 1; else; echo 1; fi;'

    at ChildProcess.exithandler (child_process.js:383:12)
    at ChildProcess.emit (events.js:400:28)
    at maybeClose (internal/child_process.js:1058:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:293:5) {
  killed: false,
  code: 1,
  signal: null,
  cmd: "ssh dani@Daniel-Mac-minije.local 'if [[ $(pmset -g powerstate IODisplayWrangler | tail -1 | cut -c29) -lt 4 ]]; then; exit 1; else; echo 1; fi;'",
  stdout: '',
  stderr: ''
}
[18/10/2021, 23:48:00] [Mac mini] Got state change from pinger for "Mac mini" (192.168.28.12): offline

I am writing this comment from my Mac mini, so it's online.
I couldn't try out the wake up or shutdown because HomeKit shows my Mac mini offline.

Is it because of M1 chip and ARM?

**Comment by AlexGustafsson** [comment]
In the terminal on your Mac, run the following commands and post the output here.

Command 1:

```shell
pmset -g powerstate IODisplayWrangler
```

Command 2:

```shell
pmset -g powerstate IODisplayWrangler | tail -1 | cut -c29
```

As you can see from your config, this command is used to check whether or not your Mac is alive. It uses the `pmset` tool to check if the display is active, which is a better indicator in some cases than ping replies. It is possible that your Mac responds with something else than the command expects. We'll hopefully be able to see that soon enough though.

Here are examples of the output to expect.

Command 1:

```
      Driver ID  Current State  Max State  Current State Description
IODisplayWrangler           4          4  USEABLE
```

Command 2:

```
4
```

**Reply by AlexGustafsson** [reply]
And to answer your question, no it's unlikely that it's due to your Mac using ARM, but there may be some other subtle differences on the new minis that no one else has stumbled across (or at least posted about here).

**Reply by duckdany** [reply]
> In the terminal on your Mac, run the following commands and post the output here.
> 
> Command 1:
> 
> ```shell
> pmset -g powerstate IODisplayWrangler
> ```
> 
> Command 2:
> 
> ```shell
> pmset -g powerstate IODisplayWrangler | tail -1 | cut -c29
> ```
> 
> As you can see from your config, this command is used to check whether or not your Mac is alive. It uses the `pmset` tool to check if the display is active, which is a better indicator in some cases than ping replies. It is possible that your Mac responds with something else than the command expects. We'll hopefully be able to see that soon enough though.
> 
> Here are examples of the output to expect.
> 
> Command 1:
> 
> ```
>       Driver ID  Current State  Max State  Current State Description
> IODisplayWrangler           4          4  USEABLE
> ```
> 
> Command 2:
> 
> ```
> 4
> ```

Command 1 output:
Driver ID  Current State  Max State  Current State Description
Internal failure: Failed to get power state information

Command 2 output:
g

Thank you!

**Reply by AlexGustafsson** [reply]
Well, there we have it. As you can see there is some issue with getting the status there, meaning the command will not work for you. Could you try the following command as well?

```shell
pmset -g powerstate | grep IO
```

On my system I get the following output:

```
IOPMrootDomain              4          4  ON
IOUSBHostDevice             2          2  USEABLE
IOBluetoothHostControlle    2          2  USEABLE
IOUSBHostDevice             2          2  USEABLE
IODisplayWrangler           4          4  USEABLE
```

**Reply by duckdany** [reply]
Output:
IOPMrootDomain              4          4  ON
IOAccessoryUSBConnectShi    1          2  USEABLE
IOThunderboltControllerT    0          0  None
IOThunderboltControllerT    0          0  None
IOThunderboltSwitchType5    0          0  None
IOThunderboltSwitchType5    0          0  None
IOThunderboltControllerT    0          0  None
IOThunderboltControllerT    0          0  None
IOThunderboltSwitchType5    0          0  None
IOThunderboltSwitchType5    0          0  None
IOUSBMassStorageInterfac    1          1  USEABLE
IOBluetoothHostControlle    1          1  USEABLE
IOUSBMassStorageDriverNu    1          1  USEABLE

Thanks

**Reply by AlexGustafsson** [reply]
Unfortunately it seems as if the previous workaround isn't usable for the new Mac mini. It seems as if the efficiency cores are so efficient, that power nap is not a thing anymore, meaning the computer is basically always on in one form or another. It may therefore respond to pings when you think it's turned off.

Currently, your best bet is to just use the ping functionality instead of the ping command. You may also want to search on the Internet for yourself for ways to tell if your Mac mini is really responding or not. Any command that may be run in the terminal will work with the plugin.

Here's one discussion about the "issue": https://forums.macrumors.com/threads/how-to-check-my-mac-mini-m1-is-really-sleeping.2284658/.

Here's information that powers the workaround for MacBooks: https://newbedev.com/macbook-does-not-sleep-when-lid-is-closed.

**Reply by dillius** [reply]
Digging around for an answer myself. Dug through all of the pmset information while SSH'd into a sleeping M1 Mac, and couldn't find any clear and obvious indicators that the machine was in a different state when comparing that output to the same machine awake.

Going to try to find some other commands outside of pmset perhaps that can be used as a test...

**Reply by dillius** [reply]
I believe I have a working solution now; gave up on SSH'ing in and checking any state, instead decided to try checking for open ports, as I know for instance things like AirPlay targets do go away when sleeping.

I had expected to find port 7001 as the AirPlay Receiver port but it didn't show as open; not sure how Apple devices actually negotiate all that, but I did discover that nmap will just fail to scan in general when the device is asleep.

So I'm using the following ping command now:
`nmap -p7000 10.0.0.116 | grep open`

**Reply by dillius** [reply]
So it turns out this check only works during the initial few seconds after sleep... not sure why. After that few seconds 7000 is responding again.

**Reply by dillius** [reply]
Next attempt appears to be working, slight modification of your original script Alex:

`ssh -i /homebridge/my-key -o StrictHostKeyChecking=no user@host 'if [[ $(pmset -g useractivity | grep Present ) ]]; then; echo 1; else; exit 1; fi;'`

**Reply by idnamrk** [reply]
Maybe this help: https://github.com/AlexGustafsson/homebridge-wol/discussions/227#discussion-4818655
